run/aem2john.py

```diff
@@ -1,74 +1,118 @@
-#
-# Output,
-#   $sspr$3$1000$a9d4b340cb43807b$33b8875ff3f9619e6ae984add262fb6b6f043e8ff9b065f4fb0863021aada275
-#   jsmith:$sspr$3$1000$fe90d85cdcd7e79c$ef182cdc47e60b472784e42a6e167d26242648c6b2e063dfd9e27eec9aa38912
-#   admin:$sspr$4$1000$fe90d85cdcd7e79c$4c29a0ac964e7bbc5380797f294d15928288cbcde3d501eb8746296de8d6c06b2b5ff27b56ae174744fe69ee157614ad126c1315ee3b67c891e42753e01a3e37
-#
-# Passwords -> admin, Aa12345678!@
+def split_hash_parts(line):
+	"""Split hash into salt, iterations, and actual hash."""
+	parts = line.split('-')
+	if len(parts) != 3:
+		return None
+	return tuple(parts)
+
+
+def process_line(line):
+	"""Process a single line from the input file."""
+	line = line.rstrip()
+	user, hash_with_algo = split_user_and_hash(line)
+	algo, hash_without_algo = extract_algo_and_hash(hash_with_algo)
+
+	if algo is None:
+		logging.warning(f"Unknown hash format in line: {line[:8]}")
+		return None
+
+	hash_parts = split_hash_parts(hash_without_algo)
+	if hash_parts is None:
+		logging.warning(f"Invalid hash format in line: {line[:8]}")
+		return None
+
+	salt, iterations, hash_value = hash_parts
+
+	return f"{user}:$sspr${algo}${iterations}${salt}${hash_value}"
 
-tag = "{SHA-256}"
-tag_length = len(tag)
 
 def process_file(filename):
-    with open(filename, "r") as f:
-        for line in f.readlines():
-            line = line.rstrip()
-            if tag in line:
-                algo = 3  # SHA-256
-            elif "{SHA-512}" in line:
-                algo = 4  # SHA-512
-            else:
-                sys.stderr.write("[!] Unknown hash format -> %s\n" % line[0:8])
-                continue
-            # Parse username
-            user = ''
-            if ':' in line:
-                parts = line.split(':')
-                user = parts[0] + ':'
-                line = parts[1]
-            # Split up hashing algo, salt, iterations, digest values
-            line = line[tag_length:]
-            data = line.split('-')
-            try:
-                salt, iterations, h = data
-            except ValueError:
-                import traceback
-                traceback.print_exc()
-                continue
-            # Print out results $$$
-            sys.stdout.write("%s$sspr$%s$%s$%s$%s\n" % (user, algo, iterations, salt, h))
+	"""Process the given input file."""
+	try:
+		with open(filename, "r") as f:
+			for line in f:
+				result = process_line(line)
+				if result is not None:
+					print(result)
+	except IOError as e:
+		logging.error(f"Error reading file {filename}: {str(e)}")
+		sys.exit(1)
 
-if __name__ == "__main__":
-    if len(sys.argv) < 2:
-        sys.stderr.write("Usage: %s <File(s)-with-Adobe-AEM-hashes>\n" % sys.argv[0])
-        sys.exit(1)
 
-    for i in range(1, len(sys.argv)):
-        process_file(sys.argv[i])
+def main():
+	parser = ArgumentParser(description="Process Adobe AEM hashes.")
```

Comment:
Maybe some more precise description like "Convert Adobe AEM hashes to the format suitable for John the Ripper"